src/garage/tf/algos/npo.py

```diff
@@ -156,13 +170,101 @@ def init_opt(self):
+            runner.step_itr += 1
+
+        return last_return
+
+    def train_once(self, itr, paths):
+        """Perform one step of policy optimization given one batch of samples.
+
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths.
+
+        Returns:
+            numpy.float64: Average return.
+
+        """
+        # -- Stage: Calculate baseline
+        paths = [
+            dict(
+                observations=self._env_spec.observation_space.flatten_n(
+                    path['observations'])
+                if self._flatten_input else path['observations'],
+                actions=(
+                    self._env_spec.action_space.flatten_n(  # noqa: E126
+                        path['actions'])),
+                rewards=path['rewards'],
+                env_infos=path['env_infos'],
+                agent_infos=path['agent_infos'],
+                dones=path['dones']) for path in paths
+        ]
+
+        if hasattr(self._baseline, 'predict_n'):
+            baseline_predictions = self._baseline.predict_n(paths)
+        else:
+            baseline_predictions = [
+                self._baseline.predict(path) for path in paths
+            ]
+
+        # -- Stage: Pre-process samples based on collected paths
+        samples_data = _functions.paths_to_tensors(paths, self.max_path_length,
+                                                   baseline_predictions,
+                                                   self._discount,
+                                                   self._gae_lambda)
+
+        # -- Stage: Run and calculate performance of the algorithm
+        undiscounted_returns = log_performance(
+            itr,
+            TrajectoryBatch.from_trajectory_list(self._env_spec, paths),
+            discount=self._discount)
+        self._episode_reward_mean.extend(undiscounted_returns)
+        tabular.record('Extras/EpisodeRewardMean',
+                       np.mean(self._episode_reward_mean))
+
+        samples_data['average_return'] = np.mean(undiscounted_returns)
+
+        self.log_diagnostics(samples_data)
+        logger.log('Optimizing policy...')
+        self.optimize_policy(itr, samples_data)
+        return samples_data['average_return']
+
+    def log_diagnostics(self, paths):
+        """Log diagnostic information.
+
+        Args:
+            paths (list[dict]): A list of collected paths.
+
+        """
+        logger.log('Logging diagnostics...')
+        self.policy.log_diagnostics(paths)
+        self._baseline.log_diagnostics(paths)
+
+    # pylint: disable=unused-argument
     def optimize_policy(self, itr, samples_data):
         """Optimize policy.
 
         Args:
             itr (int): Iteration number.
             samples_data (dict): Processed sample data.
-                See process_samples() for details.
+                See _functions.paths_to_tensors() for details.
```

Comment:
We use the _-prefixed package names to help users import functions from their canonical package paths rather than their full paths. In this case, I think that means you should point them to `garage.tf.paths_to_tensors()`